pkg/ingester/wal.go

```diff
@@ -324,33 +324,53 @@ func (w *walWrapper) performCheckpoint(immediate bool) (err error) {
 		return nil
 	}
 
-	var ticker *time.Ticker
-	if !immediate {
-		perSeriesDuration := w.cfg.CheckpointDuration / time.Duration(numSeries)
-		ticker = time.NewTicker(perSeriesDuration)
-		defer ticker.Stop()
-	}
+	perSeriesDuration := (95 * w.cfg.CheckpointDuration) / (100 * time.Duration(numSeries))
 
 	var wireChunkBuf []client.Chunk
-	b := make([]byte, 0, 1024)
+	var b []byte
+	bytePool := sync.Pool{
+		New: func() interface{} {
+			return make([]byte, 0, 1024)
+		},
+	}
+	records := [][]byte{}
+	totalSize := 0
 	for userID, state := range us {
 		for pair := range state.fpToSeries.iter() {
 			state.fpLocker.Lock(pair.fp)
-			wireChunkBuf, b, err = w.checkpointSeries(checkpoint, userID, pair.fp, pair.series, wireChunkBuf, b)
+			wireChunkBuf, b, err = w.checkpointSeries(userID, pair.fp, pair.series, wireChunkBuf, bytePool.Get().([]byte))
 			state.fpLocker.Unlock(pair.fp)
 			if err != nil {
 				return err
 			}
 
+			records = append(records, b)
+			totalSize += len(b)
+			if totalSize >= 1*1024*1024 { // 1 MiB.
+				if err := checkpoint.Log(records...); err != nil {
+					return err
+				}
+				w.checkpointLoggedBytesTotal.Add(float64(totalSize))
+				totalSize = 0
+				for i := range records {
+					bytePool.Put(records[i]) // nolint:staticcheck
+				}
+				records = records[:0]
+			}
+
 			if !immediate {
 				select {
-				case <-ticker.C:
+				case <-time.After(perSeriesDuration):
 				case <-w.quit: // When we're trying to shutdown, finish the checkpoint as fast as possible.
 				}
 			}
 		}
 	}
 
+	if err := checkpoint.Log(records...); err != nil {
```

Comment:
There may be no records at this point.

Reply:
Yes, but there maybe be records too. By doing like this we skip a `if` check as 0 records would make it a noop.